### PR TITLE
feat(static): static select support

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

static queries now support arbitrary select expressions, e.g. 

```sql
SELECT a, b + 1, subString(c) FROM X;
```

As part of this PR I've looked to start to make use of the `QueryAnalyzer` to analyze static queries.  I've also pulled the validation of static and continuous queries into their own validator classes, to make them more targeted: `StaticQueryValidator` and `ContinuousQueryValiator`, respectively.

As before, don't worry overly about the code in `StaticQueryExecutor` - it's temporary.

### Testing done 

Suitable tests added, then `mvn test`

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/test/java/io/confluent/ksql/analyzer/QueryAnalyzerTest.java [hunk 1]
```diff
@@ -15,472 +15,85 @@
...
 import org.junit.Test;
 import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.mockito.InOrder;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
 
-@SuppressWarnings("OptionalGetWithoutIsPresent")
+@RunWith(MockitoJUnitRunner.class)
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/analyzer/QueryAnalyzerTest.java]: This test was actually just renamed to `QueryAnalyzerFunctionalTest`.  This is just a bad diff by Git. 

`QueryAnalyzerFunctionalTest` continuous to be a functional style that parses actual sql and then analyses it.
`QueryAnalyzerTest` is now a unit test.

File: ksql-engine/src/test/java/io/confluent/ksql/analyzer/QueryFunctionalAnalyzerTest.java [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+
+/**
+ * DO NOT ADD NEW TESTS TO THIS FILE
+ *
+ * <p>Instead add new JSON based tests to QueryTranslationTest
+ *
+ * <p>This test file is more of a functional test, which is better implemented using QTT.
+ */
+@SuppressWarnings("OptionalGetWithoutIsPresent")
+public class QueryFunctionalAnalyzerTest {
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/analyzer/QueryFunctionalAnalyzerTest.java]: renamed from `QueryAnalyzerTest`.  Please do not review this file - I haven't changed it!

File: ksql-engine/src/test/java/io/confluent/ksql/analyzer/QueryAnalyzerFunctionalTest.java [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+
+/**
+ * DO NOT ADD NEW TESTS TO THIS FILE
+ *
+ * <p>Instead add new JSON based tests to QueryTranslationTest
+ *
+ * <p>This test file is more of a functional test, which is better implemented using QTT.
+ */
+@SuppressWarnings("OptionalGetWithoutIsPresent")
+public class QueryAnalyzerFunctionalTest {
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/analyzer/QueryAnalyzerFunctionalTest.java]: renamed from `QueryAnalyzerTest`.  Please do not review this file - I haven't changed it!

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java [hunk 4]
```diff
@@ -146,9 +146,14 @@ Analysis analyze(
     // CHECKSTYLE_RULES.ON: ClassDataAbstractionCoupling
 
     private final Analysis analysis = new Analysis();
+    private final boolean staticQuery;
     private boolean isJoin = false;
     private boolean isGroupBy = false;
 
+    Visitor(final boolean staticQuery) {
+      this.staticQuery = staticQuery;
```
Comment (purplefox) [inline ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java]: null check?

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java [hunk 5]
```diff
@@ -301,19 +306,26 @@ private void throwOnUnknownColumnReference() {
...
       for (final Expression selectExpression : analysis.getSelectExpressions()) {
-        expressionAnalyzer.analyzeExpression(selectExpression);
+        expressionAnalyzer.analyzeExpression(selectExpression, false);
       }
 
       if (analysis.getWhereExpression() != null) {
-        expressionAnalyzer.analyzeExpression(analysis.getWhereExpression());
+        final boolean allowWindowMetaFields = staticQuery
+            && analysis.getFromDataSources().get(0)
+            .getDataSource()
```
Comment (purplefox) [inline ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java]: Can we be sure there is at least one datasource? Other will NPE. Might be worth a sanity check.

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/QueryValidator.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package io.confluent.ksql.analyzer;
+
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.parser.tree.Sink;
+import java.util.Optional;
+
+/**
+ * Vaidator used by {@link QueryAnalyzer}.
```
Comment (purplefox) [inline ksql-engine/src/main/java/io/confluent/ksql/analyzer/QueryValidator.java]: typo "Vaidator".

This comment doesn't have much useful info, probably not worth adding?

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/StaticQueryValidator.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      final Query query,
+      final Optional<Sink> sink
+  ) {
+    if (!query.isStatic()) {
+      throw new IllegalArgumentException("not static");
+    }
+
+    if (query.getResultMaterialization() != ResultMaterialization.FINAL) {
+      throw new KsqlException("Static queries do not yet support `EMIT CHANGES`. "
+          + "Consider removing 'EMIT CHANGES' to any bare query, "
```
Comment (purplefox) [inline ksql-engine/src/main/java/io/confluent/ksql/analyzer/StaticQueryValidator.java]: Full stop not comma at end of line

File: ksql-engine/src/test/java/io/confluent/ksql/analyzer/ContinuousQueryValidatorTest.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  public void setUp() {
+    validator = new ContinuousQueryValidator();
+
+    when(query.isStatic()).thenReturn(false);
+  }
+
+  @Test
+  public void shouldThrowOnContinuousQueryThatIsFinal() {
+    // Given:
+    when(query.isStatic()).thenReturn(false);
```
Comment (purplefox) [inline ksql-engine/src/test/java/io/confluent/ksql/analyzer/ContinuousQueryValidatorTest.java]: redundant? same line is in setUp()

File: ksql-engine/src/test/java/io/confluent/ksql/analyzer/ExpressionAnalyzerTest.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    final Expression expression = new ComparisonExpression(
+        Type.EQUAL,
+        WINDOW_START_EXP,
+        OTHER_EXP
+    );
+
+    // When:
+    analyzer.analyzeExpression(expression, true);
+
+    // Then: did not throw
```
Comment (purplefox) [inline ksql-engine/src/test/java/io/confluent/ksql/analyzer/ExpressionAnalyzerTest.java]: The test will pass if analyzer.analyzeExpression does nothing at all. Is there some result that can be verified?

File: ksql-engine/src/test/java/io/confluent/ksql/analyzer/ExpressionAnalyzerTest.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        QualifiedName.of("fully", "qualified")
+    );
+
+    when(sourceSchemas.sourcesWithField("qualified"))
+        .thenReturn(ImmutableSet.of("multiple", "sources", "fully"));
+
+    // When:
+    analyzer.analyzeExpression(expression, true);
+
+    // Then: did not throw
```
Comment (purplefox) [inline ksql-engine/src/test/java/io/confluent/ksql/analyzer/ExpressionAnalyzerTest.java]: As above. Should verify something.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/StaticQueryExecutor.java [hunk 11]
```diff
@@ -237,7 +277,7 @@ private static WhereInfo extractWhereInfo(
 
     if (!windowed) {
       if (comparisons.size() > 1) {
-        throw invalidWhereClauseException("Unsupported WHERE clause", windowed);
+        throw invalidWhereClauseException("Unsupported WHERE clause", false);
```
Comment (purplefox) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/StaticQueryExecutor.java]: Inconsistent case compare with preceeding "missing WHERE clause". Either Unsupported and Missing or unsupported and missing.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/StaticQueryExecutor.java [hunk 12]
```diff
@@ -465,32 +505,73 @@ private static ComparisonTarget extractWhereClauseTarget(final ComparisonExpress
...
+    final LogicalSchema.Builder schema = LogicalSchema.builder();
+    schema.keyColumns(input.schema.key());
+
+    final ExpressionTypeManager expressionTypeManager = new ExpressionTypeManager(
+        input.schema,
+        executionContext.getMetaStore()
+    );
+
+    final List<SelectExpression> selects = new ArrayList<>(selectItems.size());
+    for (int idx = 0; idx < analysis.getSelectExpressions().size(); idx++) {
```
Comment (purplefox) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/StaticQueryExecutor.java]: For simple loop variables, are three letter names really necessary? i might be better

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java (hunk 4 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java]: It's a boolean... can't be null

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java (hunk 5 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java]: There's always at least one source, by definition.  In this case there is a check before to handle joins, so there must be only a single source by this point.

File: ksql-engine/src/test/java/io/confluent/ksql/analyzer/ExpressionAnalyzerTest.java (hunk 9 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/analyzer/ExpressionAnalyzerTest.java]: Nope. The only thing the class under test does is validation, i.e. it either throws or it doesn't. This test tests it doesn't throw for a specific use case

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/StaticQueryExecutor.java (hunk 12 above)
Comment (big-andy-coates) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/StaticQueryExecutor.java]: lol. super nit!  I prefer `idx` and we've no style guide that says I can't ;)

File: ksql-engine/src/test/java/io/confluent/ksql/analyzer/ContinuousQueryValidatorTest.java (hunk 8 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/analyzer/ContinuousQueryValidatorTest.java]: yup. ta.

File: ksql-engine/src/test/java/io/confluent/ksql/analyzer/ExpressionAnalyzerTest.java (hunk 10 above)
Comment (big-andy-coates) [reply ksql-engine/src/test/java/io/confluent/ksql/analyzer/ExpressionAnalyzerTest.java]: As above, nothing to verify.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/StaticQueryExecutor.java (hunk 11 above)
Comment (big-andy-coates) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/StaticQueryExecutor.java]: In some cases the user hasn't supplied a WHERE clause, so it's missing.
In other cases the user has supplied a WHERE clause but it contains things we don't support, so it's unsupported.  

These are not inconsistent... they are different!

Comment (big-andy-coates) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/StaticQueryExecutor.java]: Ah, you mean the capitalisation of the error message!

File: ksql-engine/src/main/java/io/confluent/ksql/analyzer/ExpressionAnalyzer.java [hunk 13]
```diff
@@ -138,5 +122,32 @@ public Object visitQualifiedNameReference(
       throwOnUnknownField(node.getName());
       return null;
     }
+
+    private void throwOnUnknownField(final QualifiedName name) {
+      final Set<String> sourcesWithField = sourceSchemas.sourcesWithField(name.name());
+      if (sourcesWithField.isEmpty()) {
+        if (allowWindowMetaFields && name.name().equals(SchemaUtil.WINDOWSTART_NAME)) {
```
Comment (vpapavas) [inline ksql-engine/src/main/java/io/confluent/ksql/analyzer/ExpressionAnalyzer.java]: Why do you check only for window start?

Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/analyzer/ExpressionAnalyzer.java]: Because that's the only windows meta field at the moment.

**Pull request merged**